Date: 2005-08-02
Initial Package Version: 2.9.6
Upstream Status: Not yet submitted
Origin: Miklos Vajna <vmiklos@frugalware.org>
Description: Adds support for estimating build time from SBU and counting SBU
after a build.
 etc/makepkg.conf |    3 +++
 scripts/makepkg  |   39 +++++++++++++++++++++++++++++++++++++++
 2 files changed, 42 insertions(+)
diff -Naur pacman-2.9.6.orig/etc/makepkg.conf pacman-2.9.6/etc/makepkg.conf
--- pacman-2.9.6.orig/etc/makepkg.conf	2005-08-02 23:27:19.000000000 +0200
+++ pacman-2.9.6/etc/makepkg.conf	2005-08-02 23:27:55.000000000 +0200
@@ -22,6 +22,9 @@
 # SMP Systems
 #export MAKEFLAGS="-j 2"
 
+# Build time of static binutils in seconds
+SBU="1"
+
 # Enable fakeroot for building packages as a non-root user
 export USE_FAKEROOT="y"
 
diff -Naur pacman-2.9.6.orig/scripts/makepkg pacman-2.9.6/scripts/makepkg
--- pacman-2.9.6.orig/scripts/makepkg	2005-08-02 23:27:19.000000000 +0200
+++ pacman-2.9.6/scripts/makepkg	2005-08-02 23:32:42.000000000 +0200
@@ -477,6 +477,32 @@
 	fi
 fi
 
+# estimate build time
+if grep -i -q "^# Compiling Time: [~0-9\.]\+ SBU$" $BUILDSCRIPT && \
+	[ ! -z "$SBU" ] && [ "$SBU" != "1" ]; then
+	pkgsbu=`grep -i "^# Compiling time: [~0-9\.]\+ SBU$" $BUILDSCRIPT|sed 's/^# Compiling [Tt]ime: ~*\([0-9\.]\+\) SBU$/\1/'`
+	secs=`echo "$pkgsbu*$SBU"|bc /dev/stdin|sed 's/\.[0-9]*$//'`
+	if [ $secs -ge 3600 ]; then
+		hrs=`echo $(($secs/3600))`
+		secs=`echo $(($secs%3600))`
+		mins=`echo $(($secs/60))`
+		secs=`echo $(($secs%60))`
+	elif [ $secs -ge 60 ]; then
+		mins=`echo $(($secs/60))`
+		secs=`echo $(($secs%60))`
+	fi
+	if [ ! -z "$hrs" ]; then
+		[ $hrs -gt 1 ] && hrs="$hrs hours " || hrs="$hrs hour "
+	fi
+	if [ ! -z "$mins" ]; then
+		[ $mins -gt 1 ] && mins="$mins minutes " || mins="$mins minute "
+	fi
+	if [ ! -z "$secs" ]; then
+		[ $secs -gt 1 ] && secs="$secs seconds " || secs="$secs second "
+	fi
+	msg "Estimated build time: $hrs$mins$secs"
+fi
+
 # retrieve sources
 msg "Retrieving Sources..."
 mkdir -p src
@@ -641,12 +667,25 @@
 
 # build
 msg "Starting build()..."
+stime=`date +%s`
 build 2>&1
 if [ $? -gt 0 ]; then
 	error "Build Failed.  Aborting..."
 	exit 2
 fi
 
+# count sbu
+if [ "$SBU" == "1" ]; then
+	# $SBU not yet set
+	msg "Elapsed Time: $(($(date +%s)-$stime)) seconds"
+elif [ ! -z "$SBU" ]; then
+	if [ `type -p bc` ]; then
+		msg "Elapsed Time: `echo -e "scale=2\n$(($(date +%s)-$stime))/$SBU"|bc /dev/stdin|sed 's/^\./0./'` SBU"
+	else
+		warning "The bc program is missing.  Cannot count SBU!"
+	fi
+fi
+
 # documentation
 msg "Preparing package documentation..."
 cd $startdir
